modules/social_features/social_group/src/CurrentGroupService.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+namespace Drupal\social_group;
+
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Plugin\Context\ContextProviderInterface;
+use Drupal\group\Entity\GroupInterface;
+
+/**
+ * Centralized ways to get the current group.
+ *
+ * @package Drupal\social_group.
+ */
+class CurrentGroupService {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  private EntityTypeManagerInterface $entityTypeManager;
+
+  /**
+   * The group route context.
+   *
+   * @var \Drupal\Core\Plugin\Context\ContextProviderInterface
+   */
+  private ContextProviderInterface $groupRouteContext;
+
+  /**
+   * Constructor.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager.
+   * @param \Drupal\Core\Plugin\Context\ContextProviderInterface $group_route_context
+   *   The group route context.
+   */
+  public function __construct(
+        EntityTypeManagerInterface $entity_type_manager,
+        ContextProviderInterface $group_route_context,
+    ) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->groupRouteContext = $group_route_context;
+  }
+
+  /**
+   * Get group from runtime contexts.
+   *
+   * @return \Drupal\group\Entity\GroupInterface|null
+   *   The current group or NULL.
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   */
+  public function fromRunTimeContexts(): ?GroupInterface {
+    $runtime_context = $this->groupRouteContext->getRuntimeContexts([]);
+    if (isset($runtime_context['group']) === FALSE) {
+      return NULL;
+    }
+
+    $group = $runtime_context['group']->getContextData()->getValue();
```

Comment:
I think the context class provides you with a shorter way to do this that doesn't require traversing the data itself, by calling `getContextValue` directly on the context. Since that throws an exception if a required context isn't present (and group is required) we should check for `hasContextValue` first though.

```suggestion
    if (!$runtime_context['group']->hasContextValue()) {
      return NULL;
    }
    
    return $runtime_context['group']->getContextValue();
```

Since PHPStan likely can't entirely follow the types we can add an `assert` to help show the invariant.


```suggestion
    if (!$runtime_context['group']->hasContextValue()) {
      return NULL;
    }
    
    $group = $runtime_context['group']->getContextValue();
    assert($group instanceof GroupInterface, "The group context resolver returned a context value that is not a GroupInterface instance which violates the services contract".);
    return $group;
```